dart/dynamics/InverseKinematics.cpp

```diff
@@ -0,0 +1,1202 @@
+  const Eigen::Isometry3d& targetTf =
+      mIK->getTarget()->getRelativeTransform();
+  // Use the actual transform with respect to the target's reference frame
+  const Eigen::Isometry3d& actualTf =
+      mIK->getNode()->getTransform(mIK->getTarget()->getParentFrame());
+
+  // ^ This scheme makes it so that the bounds are expressed in the reference
+  // frame of the target
+
+  Eigen::Vector3d p_error = actualTf.translation() - targetTf.translation();
+  if(mIK->hasOffset())
+    p_error += actualTf.linear()*mIK->getOffset();
+
+  Eigen::Matrix3d R_error = actualTf.linear() * targetTf.linear().transpose();
+
+  Eigen::Vector6d displacement;
+  displacement.head<3>() = math::matrixToEulerXYZ(R_error);
+  displacement.tail<3>() = p_error;
+
+  Eigen::Vector6d error;
+  const Eigen::Vector6d& min = mProperties.mBounds.first;
+  const Eigen::Vector6d& max = mProperties.mBounds.second;
+  double tolerance = mIK->getSolver()->getTolerance();
+  for(int i=0; i<6; ++i)
+  {
+    if( displacement[i] < min[i] )
+    {
+      if(mComputeErrorFromCenter)
+      {
+        if(!std::isinf(max[i]))
+          error[i] = displacement[i] - (min[i]+max[i])/2.0;
+        else
+          error[i] = displacement[i] - (min[i]+tolerance);
+      }
+      else
+      {
+        error[i] = displacement[i] - min[i];
+      }
+    }
+    else if( max[i] < displacement[i] )
+    {
+      if(mComputeErrorFromCenter && !std::isinf(min[i]))
+      {
+        if(!std::isinf(min[i]))
+          error[i] = displacement[i] - (min[i]+max[i])/2.0;
+        else
+          error[i] = displacement[i] - (max[i]-tolerance);
+      }
+      else
+      {
+        error[i] = displacement[i] - max[i];
+      }
+    }
+    else
+      error[i] = 0.0;
+  }
+
+  error = error.cwiseProduct(mProperties.mErrorWeights);
+
+  if(error.norm() > mProperties.mErrorLengthClamp)
+    error = error.normalized()*mProperties.mErrorLengthClamp;
+
+  if(!mIK->getTarget()->getParentFrame()->isWorld())
+  {
+    // Transform the error term into the world frame if it's not already
+    const Eigen::Isometry3d& R =
+        mIK->getTarget()->getParentFrame()->getWorldTransform();
+    error.head<3>() = R.linear()*error.head<3>();
+    error.tail<3>() = R.linear()*error.tail<3>();
+  }
+
+  return error;
+}
+
+//==============================================================================
+InverseKinematics::GradientMethod::GradientMethod(
+    InverseKinematics* _ik,
+    const std::string& _methodName, double _clamp)
+  : mIK(_ik),
+    mMethodName(_methodName),
+    mComponentWiseClamp(_clamp)
+{
+  // Do nothing
+}
+
+//==============================================================================
+void InverseKinematics::GradientMethod::evalGradient(
+    const Eigen::VectorXd& _q,
+    Eigen::Map<Eigen::VectorXd> _grad)
+{
+  if(_q.size() != static_cast<int>(mIK->getDofs().size()))
+  {
+    dterr << "[InverseKinematics::GradientMethod::computeGradient] Mismatch "
```

Comment:
I found that there is a standard predefined macro for function name, `__func__`, but it seems there is no one for class name even of non standard macros.